File: src/lowlevel/server/icefsoperators/icefsCopyFileRange.go
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-11 07:18:32
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-17 15:11:11
 * @FilePath: /icefs/src/lowlevel/server/icefsoperators/icefsCopyFileRange.go
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
package icefsoperators

import (
	"context"
	"icefs-server/icefserror"
	pb "icefs-server/icefsgrpc"
	"icefs-server/icefsthrift"

	"golang.org/x/sys/unix"
)

func (s *IcefsServer) doIcefsCopyFileRange(offsetIn int64, fhIn uint64, offsetOut int64, fhOut uint64, len uint64, flags int32) (status int32, size uint64) {
	num, err := unix.CopyFileRange(int(fhIn), &offsetIn, int(fhOut), &offsetOut, int(len), int(flags))
	status = icefserror.IcefsStdErrno(err)
	size = uint64(num) // 保证先进行status校验则回环不会出现错误
	return
}

func (s *IcefsGRpcServer) DoIcefsCopyFileRange(ctx context.Context, req *pb.IcefsCopyFileRangeReq) (*pb.IcefsCopyFileRangeRes, error) {
	var res pb.IcefsCopyFileRangeRes

	res.Status, res.Size = s.server.doIcefsCopyFileRange(req.OffsetIn, req.FhIn, req.OffsetOut, req.FhOut, req.Len, req.Flags)

	return &res, nil
}

func (s *IcefsThriftServer) DoIcefsCopyFileRange(ctx context.Context, req *icefsthrift.IcefsCopyFileRangeReq) (*icefsthrift.IcefsCopyFileRangeRes, error) {
	var res icefsthrift.IcefsCopyFileRangeRes
	var size uint64

	res.Status, size = s.server.doIcefsCopyFileRange(req.OffsetIn, uint64(req.FhIn), req.OffsetOut, uint64(req.FhOut), uint64(req.Len), req.Flags)
	res.Size = icefsthrift.Ui64(size)

	return &res, nil
}

File: src/lowlevel/client/icefsoperators/icefsCopyFileRange.cpp
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-30 04:19:29
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-21 02:37:49
 * @FilePath: /icefs/src/lowlevel/client/icefsoperators/icefsCopyFileRange.cpp
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
#include <stdio.h>

#include <string>

#include "icefsClient.hpp"
#include "icefsClientUtils.hpp"

void IcefsClient::DoIcefsCopyFileRange(fuse_req_t fuseReq, fuse_ino_t inodeIn,
                                       off_t offsetIn,
                                       struct fuse_file_info *fiIn,
                                       fuse_ino_t inodeOut, off_t offsetOut,
                                       struct fuse_file_info *fiOut, size_t len,
                                       int flags) {
  ICEFS_PR_FUNCTION;

  switch (this->clientConfig.linkType) {
    case ICEFS_LINK_USE_GRPC: {
      icefsgrpc::IcefsCopyFileRangeReq req;
      icefsgrpc::IcefsCopyFileRangeRes res;
      grpc::ClientContext ctx;

      req.set_offset_in(offsetIn);
      req.set_fh_in(fiIn->fh);
      req.set_offset_out(offsetOut);
      req.set_fh_out(fiOut->fh);
      req.set_len(len);
      req.set_flags(flags);

      grpc::Status status = gRpcClient->DoIcefsCopyFileRange(&ctx, req, &res);
      if (status.ok() && !res.status()) {
        fuse_reply_write(fuseReq, res.size());
      } else {
        fuse_reply_err(fuseReq, res.status() ? res.status() : EIO);
        ICEFS_PR_ERR_STATUS;
      }

      break;
    }

    case ICEFS_LINK_USE_THRIFT: {
      icefsthrift::IcefsCopyFileRangeReq req;
      icefsthrift::IcefsCopyFileRangeRes res;

      req.__set_offset_in(offsetIn);
      req.__set_fh_in(fiIn->fh);
      req.__set_offset_out(offsetOut);
      req.__set_fh_out(fiOut->fh);
      req.__set_len(len);
      req.__set_flags(flags);

      icefsThriftConn *thriftConn = thriftClientPool->GetIcefsThriftConn();
thriftConn->thriftClient->DoIcefsCopyFileRange(res, req);
thriftClientPool->PutIcefsThriftConn(thriftConn);
      if (!res.status) {
        fuse_reply_write(fuseReq, res.size);
      } else {
        fuse_reply_err(fuseReq, res.status);
        ICEFS_PR_ERR_STATUS;
      }
      break;
    }

    default:
      fuse_reply_err(fuseReq, EIO);
      break;
  }
}